File: 프로그래머스/카카오/[3차] 압축.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
#include <map>

using namespace std;

queue<char> q;
map<string, int> mp;

vector<int> solution(string msg) {
    vector<int> answer;
    for (int i = 0; i < 26; ++i) {
        char x = 'A' + i;
        string s(1,x);
        mp[s] = i+1;
    }

    for (char ch : msg) {
        q.push(ch);
    }

    int index = 27;
    string tmp="";
    while (!q.empty()) {
        char ch = q.front();
        if (mp[tmp+ch]) {//사전에 있다면
            tmp+=ch;
            q.pop();
            if (q.empty()) {
                answer.push_back(mp[tmp]);
            }
            continue;
        }else{
            mp[tmp+ch] = index++;
            answer.push_back(mp[tmp]);
            tmp="";
        }
    }
    return answer;
}

File: 프로그래머스/카카오/[3차] 압축.py
def solution(msg):
    answer = []
    dic = dict()
    for i in range(26):
        dic[chr(65 + i)] = i + 1

    value = 27
    s = 0
    e = 1
    while e <= len(msg):
        tmp = msg[s:e]
        if not tmp in dic:
            answer.append(dic[msg[s:e-1]])
            dic[tmp] = value
            value+=1
            s= e-1
            e=s
        e+=1
    answer.append(dic[msg[s:e-1]])

    return answer
